cadc-dali/src/main/java/ca/nrc/cadc/dali/tables/parquet/ParquetWriter.java

```diff
@@ -0,0 +1,306 @@
+                            throw new RuntimeException(e);
+                        }
+                    }
+
+                    @Override
+                    public long getPos() {
+                        return position;
+                    }
+
+                    @Override
+                    public void close() {
+                        try {
+                            outputStream.close();
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                };
+            }
+
+            @Override
+            public PositionOutputStream createOrOverwrite(long blockSizeHint) {
+                return create(blockSizeHint);
+            }
+
+            @Override
+            public boolean supportsBlockSize() {
+                return false;
+            }
+
+            @Override
+            public long defaultBlockSize() {
+                return 0;
+            }
+        };
+    }
+
+    private static void updateVOTable(VOTableResource voTableResource) {
+        List<VOTableField> fields = voTableResource.getTable().getFields();
+
+        // Convert timestamp fields to long to ensure inconsistencies between avro schema field and votable metadata field
+        for (int i = 0; i < fields.size(); i++) {
+            VOTableField field = fields.get(i);
+
+            if ("timestamp".equals(field.xtype)) {
+                fields.set(i, new VOTableField(field.getName(), "long"));
+            }
+
+            if ("short".equals(field.getDatatype())) {
+                fields.set(i, new VOTableField(field.getName(), "int", field.getArraysize()));
+            }
+        }
+
+        // Clear table data for empty VOTable
+        voTableResource.getTable().setTableData(null);
+    }
+
+    private static Map<String, String> prepareCustomMetaData(VOTableDocument voTableDocument, Long maxRec) throws IOException {
+        StringWriter stringWriter = new StringWriter();
+        VOTableWriter votableWriter = new VOTableWriter();
+        votableWriter.write(voTableDocument, stringWriter, maxRec);
+
+        Map<String, String> customMetaData = new HashMap<>();
+        customMetaData.put(IVOA_VOTABLE_PARQUET_VERSION_KEY, IVOA_VOTABLE_PARQUET_VERSION_VALUE);
+        customMetaData.put(IVOA_VOTABLE_PARQUET_CONTENT_KEY, stringWriter.toString());
+
+        return customMetaData;
+    }
+
+    private static int saveRecords(Long maxRec, TableData tableData, Schema schema, org.apache.parquet.hadoop.ParquetWriter<GenericRecord> writer)
+            throws IOException {
+        Iterator<List<Object>> iterator = tableData.iterator();
+        int recordCount = 1;
+
+        while (iterator.hasNext() && recordCount <= maxRec) {
+            GenericRecord record = new GenericData.Record(schema);
+            List<Object> rowData = iterator.next();
+
+            for (Schema.Field field : schema.getFields()) {
```

Comment:
I would prefer this `for` loop be over the rowData list so this is more clearly "loop over the data", but then you need tom be able to get the field from the schema by position... don't know if that's supported.